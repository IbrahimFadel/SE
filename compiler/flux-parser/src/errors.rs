use ariadne::Report;
use flux_lexer::TokenKind;
use flux_span::{Span, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub enum ParseError {
	ExpectedToken {
		expected: Vec<TokenKind>,
		got: Spanned<TokenKind>,
	},
	// More general which can allow for more specific messages and for expecting larger constructs like types or parameter
	Expected {
		expected: Spanned<String>,
	},
}

impl<'a> Into<&'a Report<Span>> for &'a ParseError {
	fn into(self) -> &'a Report<Span> {
		todo!()
	}
}
