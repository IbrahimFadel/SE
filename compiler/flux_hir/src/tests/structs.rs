use crate::errors;

errors! {
    field_type_mismatches,
    r#"
//- main.flx

struct Foo {
    x s32,
    y Bar,
    z f32,
}

struct Bar {
    x u32
}

fn main() {
    let foo_instance = Foo {
        x: 0.0,
        y: 10,
        z: 1,
    };
}
"#
}

errors! {
    uninitialized_fields,
    r#"
//- main.flx

struct Foo {
    x s32,
    y Bar,
    z f32,
}

struct Bar {
    x u32
}

fn main() {
    let foo_instance = Foo {
        x: 1,
    };
}
"#
}

errors! {
    unknown_fields,
    r#"
//- main.flx

struct Foo {
    x s32,
    y Bar,
    z f32,
}

struct Bar {
    x u32
}

fn main() {
    let foo_instance = Foo {
        x: 0,
        foo: 0,
        bar: 0,
        bazz: 0,
    };
}
"#
}

errors! {
    typecheck_field_access,
    r#"
//- main.flx

struct Foo {
    x u32,
}

fn main() {
    let foo = Foo {
        x: 0
    };

    let x s32 = foo.x;
}
"#
}

errors! {
    typecheck_method_call,
    r#"
//- main.flx

struct Foo {}

apply to Foo {
    fn foo() {}
}

fn main() {
    let foo = Foo {};

    let x s32 = foo.foo();
}
"#
}

errors! {
    unknown_method_called,
    r#"
//- main.flx

struct Foo {}

fn main() {
    let foo = Foo {};

    foo.foo();
}
"#
}

errors! {
    unknown_field_referenced,
    r#"
//- main.flx

struct Foo {}

fn main() {
    let foo = Foo {};

    let x = foo.x;
}
"#
}

errors! {
    generic_typecheck_in_initialization,
    r#"
//- main.flx

trait Bar {}

struct Foo<T>
    where T is Bar
{
    x T
}

struct Bazz {}

fn main() {
    let bazz = Bazz {};
    let foo = Foo {
        x: bazz,
    };
}
"#
}
