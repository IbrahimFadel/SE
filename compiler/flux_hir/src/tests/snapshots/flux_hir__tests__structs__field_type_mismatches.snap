---
source: compiler/flux_hir/src/tests/structs.rs
expression: s
---
//  main.flx

struct Foo {
s32 x,Bar y,f32 z}

struct Bar {
u32 x}

fn main() -> () {
    let foo_instance Foo = struct Foo {
        x: 0,  y: 10,  z: 1
    };
}

Diagnostics

[E0041] Error: type mismatch
    ,-[ main.flx:14:24]
    |
  4 |     x s32,
    |    ^|^  
    |     `--- `s32`
    | 
 14 |     let foo_instance = Foo {
    |                        |  ^|^  
    |                        `------- type mismatch between `s32` and `float`
    |                            |   
    |                            `--- `float`
----'
[E0041] Error: type mismatch
    ,-[ main.flx:15:10]
    |
  5 |     y Bar,
    |   ^|^  
    |    `--- `Bar`
    | 
 15 |         x: 0.0,
    |          |  ^|  
    |          `------ type mismatch between `Bar` and `int`
    |              |  
    |              `-- `int`
----'
[E0041] Error: type mismatch
    ,-[ main.flx:16:8]
    |
  6 |     z f32,
    |  ^|^  
    |   `--- `f32`
    | 
 16 |         y: 10,
    |        |  |  
    |        `----- type mismatch between `f32` and `int`
    |           |  
    |           `-- `int`
----'

